src/python/pants/backend/python/util_rules/dists.py

```diff
@@ -0,0 +1,233 @@
+        build_wheel=request.build_wheel,
+        build_sdist=request.build_sdist,
+        config_settings_str=distutils_repr(request.config_settings),
+    ).encode()
+
+
+@rule
+async def run_pep517_build(request: DistBuildRequest, python_setup: PythonSetup) -> DistBuildResult:
+    # Note that this pex has no entrypoint. We use it to run our generated shim, which
+    # in turn imports from and invokes the build backend.
+    build_backend_pex = await Get(
+        VenvPex,
+        PexRequest(
+            output_filename="build_backend.pex",
+            internal_only=True,
+            requirements=request.requires,
+            interpreter_constraints=request.interpreter_constraints,
+        ),
+    )
+
+    dist_dir = "dist"
+    backend_shim_name = "backend_shim.py"
+    backend_shim_path = os.path.join(request.working_directory, backend_shim_name)
+    backend_shim_digest = await Get(
+        Digest,
+        CreateDigest(
+            [
+                FileContent(backend_shim_path, interpolate_backend_shim(dist_dir, request)),
+            ]
+        ),
+    )
+
+    merged_digest = await Get(Digest, MergeDigests((request.input, backend_shim_digest)))
+
+    if python_setup.macos_big_sur_compatibility and is_macos_big_sur():
+        extra_env = {"MACOSX_DEPLOYMENT_TARGET": "10.16"}
+    else:
+        extra_env = {}
+
+    result = await Get(
+        ProcessResult,
+        VenvPexProcess(
+            build_backend_pex,
+            argv=(backend_shim_name,),
+            input_digest=merged_digest,
+            extra_env=extra_env,
+            working_directory=request.working_directory,
+            output_directories=(dist_dir,),  # Relative to the working_directory.
+            description=(
+                f"Run {request.build_backend} for {request.target_address_spec}"
+                if request.target_address_spec
+                else f"Run {request.build_backend}"
+            ),
+            level=LogLevel.DEBUG,
+        ),
+    )
+    output_lines = result.stdout.decode().splitlines()
+    paths = {}
+    for line in output_lines:
+        for dist_type in ["wheel", "sdist"]:
+            if line.startswith("{}: ".format(dist_type)):
+                paths[dist_type] = line[len(dist_type) + 2 :].strip()
+    # Note that output_digest paths are relative to the working_directory.
+    output_digest = await Get(Digest, RemovePrefix(result.output_digest, dist_dir))
+    output_snapshot = await Get(Snapshot, Digest, output_digest)
+    for dist_type, path in paths.items():
+        if path not in output_snapshot.files:
+            raise BuildBackendError(
+                f"Build backend {request.build_backend} did not create "
+                f"expected {dist_type} file {path}"
+            )
+    return DistBuildResult(
+        output_digest, wheel_path=paths.get("wheel"), sdist_path=paths.get("sdist")
+    )
+
+
+# Distutils does not support unicode strings in setup.py, so we must explicitly convert to binary
```

Comment:
It might be good to experiment with dropping this together with a py27 build with config_settings. Since there is no more setup.py generated by us, we may not need a this hackery any longer.

Reply:
Well, we're still generating setup.py for now, but even when we don't don't, we still need config_settings in the shim to be py27-compatible, no? What is difference about this? (I don't totally know why we ever needed it).

Reply:
Ah, right. We shouldn't be generating setup.py - except for `from setuptools import setup; setup()`. Instead we should dump all config to `setup.cfg` since modern setuptools supports this if just to highlight we don't do work in `setup.py`. That's for another change though.

As to the reason for all this, it's in the comment. Something deep in distutils would blow up on unicode literals. Presumably this is handled though by setuptools when it translates setup.cfg to arguments to distutils under the covers

Reply:
An example of pure declarative for setuptools: https://github.com/pantsbuild/pex/blob/d6235f55282521d28448422d773ec58280f85fdc/pex/tools/commands/repository.py#L353

Reply:
Hrm. Right. I guess we were paying an encoding price for getting away with not having to understand setup kwargs. The approach I point out does understand kwargs which is the key thing. Humbug.

Reply:
Oh, hm, my point was more that I we are now using this helper to generate the `config_settings` literal in this new shim. But since the issue was distutils-specific then I guess this is unnecessary? Also, we do in fact know the types in config_settings. We have declared it by fiat to be string-to-list-of-strings. So maybe we can indeed do something much simpler (once we stop generating setup.py).

Reply:
Yeah, since you've pinned down the types, this can be made simpler in a follow-up.